File: core/src/main/java/com/shatteredpixel/shatteredpixeldungeon/levels/rooms/Room.java
/*
 * Pixel Dungeon
 * Copyright (C) 2012-2015 Oleg Dolya
 *
 * Shattered Pixel Dungeon
 * Copyright (C) 2014-2019 Evan Debenham
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */

package com.shatteredpixel.shatteredpixeldungeon.levels.rooms;

import com.shatteredpixel.shatteredpixeldungeon.levels.Level;
import com.watabou.utils.Bundlable;
import com.watabou.utils.Bundle;
import com.watabou.utils.Graph;
import com.watabou.utils.Point;
import com.watabou.utils.Random;
import com.watabou.utils.Rect;

import java.util.ArrayList;
import java.util.Collection;
import java.util.LinkedHashMap;

public abstract class Room extends Rect implements Graph.Node, Bundlable {
	
	public ArrayList<Room> neigbours = new ArrayList<>();
	public LinkedHashMap<Room, Door> connected = new LinkedHashMap<>();
	
	public int distance;
	public int price = 1;
	
	public Room(){
		super();
	}
	
	public Room( Rect other ){
		super(other);
	}
	
	public Room set( Room other ) {
		super.set( other );
		for (Room r : other.neigbours){
			neigbours.add(r);
			r.neigbours.remove(other);
			r.neigbours.add(this);
		}
		for (Room r : other.connected.keySet()){
			Door d = other.connected.get(r);
			r.connected.remove(other);
			r.connected.put(this, d);
			connected.put(r, d);
		}
		return this;
	}
	
	// **** Spatial logic ****
	
	//Note: when overriding these YOU MUST store any randomly decided values.
	//With the same room and the same parameters these should always return
	//the same value over multiple calls, even if there's some randomness initially.
	public int minWidth(){
		return -1;
	}
	public int maxWidth() { return -1; }
	
	public int minHeight() { return -1; }
	public int maxHeight() { return -1; }
	
	public boolean setSize(){
		return setSize(minWidth(), maxWidth(), minHeight(), maxHeight());
	}
	
	public boolean forceSize( int w, int h ){
		return setSize( w, w, h, h );
	}
	
	public boolean setSizeWithLimit( int w, int h ){
		if ( w < minWidth() || h < minHeight()) {
			return false;
		} else {
			setSize();
			
			if (width() > w || height() > h){
				resize(Math.min(width(), w)-1, Math.min(height(), h)-1);
			}
			
			return true;
		}
	}
	
	protected boolean setSize(int minW, int maxW, int minH, int maxH) {
		if (minW < minWidth()
				|| maxW > maxWidth()
				|| minH < minHeight()
				|| maxH > maxHeight()
				|| minW > maxW
				|| minH > maxH){
			return false;
		} else {
			//subtract one because rooms are inclusive to their right and bottom sides
			resize(Random.NormalIntRange(minW, maxW) - 1,
					Random.NormalIntRange(minH, maxH) - 1);
			return true;
		}
	}
	
	//Width and height are increased by 1 because rooms are inclusive to their right and bottom sides
	@Override
	public int width() {
		return super.width()+1;
	}
	
	@Override
	public int height() {
		return super.height()+1;
	}
	
	public Point random() {
		return random( 1 );
	}
	
	public Point random( int m ) {
		return new Point( Random.IntRange( left + m, right - m ),
				Random.IntRange( top + m, bottom - m ));
	}
	
	//a point is only considered to be inside if it is within the 1 tile perimeter
	public boolean inside( Point p ) {
		return p.x > left && p.y > top && p.x < right && p.y < bottom;
	}
	
	public Point center() {
		return new Point(
				(left + right) / 2 + (((right - left) % 2) == 1 ? Random.Int( 2 ) : 0),
				(top + bottom) / 2 + (((bottom - top) % 2) == 1 ? Random.Int( 2 ) : 0) );
	}
	
	
	// **** Connection logic ****
	
	public static final int ALL     = 0;
	public static final int LEFT    = 1;
	public static final int TOP     = 2;
	public static final int RIGHT   = 3;
	public static final int BOTTOM  = 4;
	
	public int minConnections(int direction){
		if (direction == ALL)   return 1;
		else                    return 0;
	}
	
	public int curConnections(int direction){
		if (direction == ALL) {
			return connected.size();
			
		} else {
			int total = 0;
			for (Room r : connected.keySet()){
				Rect i = intersect( r );
				if      (direction == LEFT && i.width() == 0 && i.left == left)         total++;
				else if (direction == TOP && i.height() == 0 && i.top == top)           total++;
				else if (direction == RIGHT && i.width() == 0 && i.right == right)      total++;
				else if (direction == BOTTOM && i.height() == 0 && i.bottom == bottom)  total++;
			}
			return total;
		}
	}
	
	public int remConnections(int direction){
		if (curConnections(ALL) >= maxConnections(ALL)) return 0;
		else return maxConnections(direction) - curConnections(direction);
	}
	
	public int maxConnections(int direction){
		if (direction == ALL)   return 16;
		else                    return 4;
	}
	
	//only considers point-specific limits, not direction limits
	public boolean canConnect(Point p){
		//point must be along exactly one edge, no corners.
		return (p.x == left || p.x == right) != (p.y == top || p.y == bottom);
	}
	
	//only considers direction limits, not point-specific limits
	public boolean canConnect(int direction){
		return remConnections(direction) > 0;
	}
	
	//considers both direction and point limits
	public boolean canConnect( Room r ){
		Rect i = intersect( r );
		
		boolean foundPoint = false;
		for (Point p : i.getPoints()){
			if (canConnect(p) && r.canConnect(p)){
				foundPoint = true;
				break;
			}
		}
		if (!foundPoint) return false;
		
		if (i.width() == 0 && i.left == left)
			return canConnect(LEFT) && r.canConnect(LEFT);
		else if (i.height() == 0 && i.top == top)
			return canConnect(TOP) && r.canConnect(TOP);
		else if (i.width() == 0 && i.right == right)
			return canConnect(RIGHT) && r.canConnect(RIGHT);
		else if (i.height() == 0 && i.bottom == bottom)
			return canConnect(BOTTOM) && r.canConnect(BOTTOM);
		else
			return false;
	}
	
	public boolean addNeigbour( Room other ) {
		if (neigbours.contains(other))
			return true;
		
		Rect i = intersect( other );
		if ((i.width() == 0 && i.height() >= 2) ||
			(i.height() == 0 && i.width() >= 2)) {
			neigbours.add( other );
			other.neigbours.add( this );
			return true;
		}
		return false;
	}
	
	public boolean connect( Room room ) {
		if ((neigbours.contains(room) || addNeigbour(room))
				&& !connected.containsKey( room ) && canConnect(room)) {
			connected.put( room, null );
			room.connected.put( this, null );
			return true;
		}
		return false;
	}
	
	public void clearConnections(){
		for (Room r : neigbours){
			r.neigbours.remove(this);
		}
		neigbours.clear();
		for (Room r : connected.keySet()){
			r.connected.remove(this);
		}
		connected.clear();
	}
	
	// **** Painter Logic ****
	
	public abstract void paint(Level level);
	
	//whether or not a painter can make its own modifications to a specific point
	public boolean canPlaceWater(Point p){
		return inside(p);
	}
	
	public final ArrayList<Point> waterPlaceablePoints(){
		ArrayList<Point> points = new ArrayList<>();
		for (int i = left; i <= right; i++) {
			for (int j = top; j <= bottom; j++) {
				Point p = new Point(i, j);
				if (canPlaceWater(p)) points.add(p);
			}
		}
		return points;
	}

	//whether or not a painter can make place grass at a specific point
	public boolean canPlaceGrass(Point p){
		return inside(p);
	}

	public final ArrayList<Point> grassPlaceablePoints(){
		ArrayList<Point> points = new ArrayList<>();
		for (int i = left; i <= right; i++) {
			for (int j = top; j <= bottom; j++) {
				Point p = new Point(i, j);
				if (canPlaceGrass(p)) points.add(p);
			}
		}
		return points;
	}
	
	//whether or not a painter can place a trap at a specific point
	public boolean canPlaceTrap(Point p){
		return inside(p);
	}
	
	public final ArrayList<Point> trapPlaceablePoints(){
		ArrayList<Point> points = new ArrayList<>();
		for (int i = left; i <= right; i++) {
			for (int j = top; j <= bottom; j++) {
				Point p = new Point(i, j);
				if (canPlaceTrap(p)) points.add(p);
			}
		}
		return points;
	}
	
	//whether or not a character (usually spawned) can be placed here
	public boolean canPlaceCharacter(Point p, Level l){
		return inside(p);
	}
	
	public final ArrayList<Point> charPlaceablePoints(Level l){
		ArrayList<Point> points = new ArrayList<>();
		for (int i = left; i <= right; i++) {
			for (int j = top; j <= bottom; j++) {
				Point p = new Point(i, j);
				if (canPlaceCharacter(p, l)) points.add(p);
			}
		}
		return points;
	}
	
	// **** Graph.Node interface ****

	@Override
	public int distance() {
		return distance;
	}

	@Override
	public void distance( int value ) {
		distance = value;
	}
	
	@Override
	public int price() {
		return price;
	}

	@Override
	public void price( int value ) {
		price = value;
	}

	@Override
	public Collection<Room> edges() {
		ArrayList<Room> edges = new ArrayList<>();
		for( Room r : connected.keySet()){
			Door d = connected.get(r);
			//for the purposes of path building, ignore all doors that are locked, blocked, or hidden
			if (d.type == Door.Type.EMPTY || d.type == Door.Type.TUNNEL
					|| d.type == Door.Type.UNLOCKED || d.type == Door.Type.REGULAR){
				edges.add(r);
			}
		}
		return edges;
	}
	
	@Override
	public void storeInBundle( Bundle bundle ) {
		bundle.put( "left", left );
		bundle.put( "top", top );
		bundle.put( "right", right );
		bundle.put( "bottom", bottom );
	}
	
	@Override
	public void restoreFromBundle( Bundle bundle ) {
		left = bundle.getInt( "left" );
		top = bundle.getInt( "top" );
		right = bundle.getInt( "right" );
		bottom = bundle.getInt( "bottom" );
	}

	//FIXME currently connections and neighbours are not preserved on load
	public void onLevelLoad( Level level ){
		//does nothing by default
	}
	
	public static class Door extends Point implements Bundlable {
		
		public enum Type {
			EMPTY, TUNNEL, REGULAR, UNLOCKED, HIDDEN, BARRICADE, LOCKED
		}
		public Type type = Type.EMPTY;
		
		public Door(){
		}
		
		public Door( Point p ){
			super(p);
		}
		
		public Door( int x, int y ) {
			super( x, y );
		}
		
		public void set( Type type ) {
			if (type.compareTo( this.type ) > 0) {
				this.type = type;
			}
		}
		
		@Override
		public void storeInBundle(Bundle bundle) {
			bundle.put("x", x);
			bundle.put("y", y);
			bundle.put("type", type);
		}
		
		@Override
		public void restoreFromBundle(Bundle bundle) {
			x = bundle.getInt("x");
			y = bundle.getInt("y");
			type = bundle.getEnum("type", Type.class);
		}
	}
}

File: pixel-dungeon-web/src/levels/rooms/Room.js
/*
 * Pixel Dungeon Web Port
 * Ported from Shattered Pixel Dungeon
 * Copyright (C) 2012-2015 Oleg Dolya
 * Copyright (C) 2014-2025 Evan Debenham
 *
 * Base Room Class with Connection Logic
 * Source: core/src/main/java/com/shatteredpixel/shatteredpixeldungeon/levels/rooms/Room.java
 */

import { Rect } from '../../utils/Rect.js';
import { Point } from '../../utils/Point.js';
import { Random } from '../../utils/Random.js';

/**
 * Door types
 * Source: Room.java:444-446
 */
export const DoorType = {
    EMPTY: 'EMPTY',
    TUNNEL: 'TUNNEL',
    WATER: 'WATER',
    REGULAR: 'REGULAR',
    UNLOCKED: 'UNLOCKED',
    HIDDEN: 'HIDDEN',
    BARRICADE: 'BARRICADE',
    LOCKED: 'LOCKED',
    CRYSTAL: 'CRYSTAL',
    WALL: 'WALL'
};

/**
 * Door class
 * Source: Room.java:442-487
 */
export class Door extends Point {
    constructor(x = 0, y = 0) {
        if (typeof x === 'object' && x instanceof Point) {
            super(x.x, x.y);
        } else {
            super(x, y);
        }
        this.type = DoorType.EMPTY;
        this.typeLocked = false;
    }

    lockTypeChanges(lock) {
        this.typeLocked = lock;
    }

    /**
     * Set door type (only if new type is "greater" than current)
     * Source: Room.java:466-470
     */
    set(type) {
        const typeOrder = Object.keys(DoorType);
        if (!this.typeLocked && typeOrder.indexOf(type) > typeOrder.indexOf(this.type)) {
            this.type = type;
        }
    }
}

/**
 * Room direction constants
 * Source: Room.java:168-172
 */
export const RoomDirection = {
    ALL: 0,
    LEFT: 1,
    TOP: 2,
    RIGHT: 3,
    BOTTOM: 4
};

/**
 * Base Room class
 * Extends Rect with graph node capabilities and connection logic
 * Source: Room.java:37-488
 */
export class Room extends Rect {
    constructor(...args) {
        super(...args);

        // Graph.Node interface
        // Source: Room.java:39-43
        this.neigbours = []; // Spelling intentional - matches Java typo
        this.connected = new Map(); // Map<Room, Door>
        this.distance = 0;
        this.price = 1;
    }

    /**
     * Set this room from another, transferring all connections
     * Source: Room.java:53-67
     */
    set_room(other) {
        super.set(other);

        // Transfer neighbors
        for (const r of other.neigbours) {
            this.neigbours.push(r);
            r.neigbours = r.neigbours.filter(n => n !== other);
            r.neigbours.push(this);
        }

        // Transfer connections
        for (const [r, d] of other.connected.entries()) {
            r.connected.delete(other);
            r.connected.set(this, d);
            this.connected.set(r, d);
        }

        return this;
    }

    // ========== Spatial Logic ==========
    // Source: Room.java:69-163

    minWidth() { return -1; }
    maxWidth() { return -1; }
    minHeight() { return -1; }
    maxHeight() { return -1; }

    setSize() {
        return this.setSize_minmax(this.minWidth(), this.maxWidth(), this.minHeight(), this.maxHeight());
    }

    forceSize(w, h) {
        return this.setSize_minmax(w, w, h, h);
    }

    /**
     * Set size with limit
     * Source: Room.java:90-102
     */
    setSizeWithLimit(w, h) {
        if (w < this.minWidth() || h < this.minHeight()) {
            return false;
        } else {
            this.setSize();

            if (this.width() > w || this.height() > h) {
                this.resize(Math.min(this.width(), w) - 1, Math.min(this.height(), h) - 1);
            }

            return true;
        }
    }

    /**
     * Set size within min/max bounds
     * Source: Room.java:104-118
     */
    setSize_minmax(minW, maxW, minH, maxH) {
        if (minW < this.minWidth()
                || maxW > this.maxWidth()
                || minH < this.minHeight()
                || maxH > this.maxHeight()
                || minW > maxW
                || minH > maxH) {
            return false;
        } else {
            // Subtract one because rooms are inclusive to their right and bottom sides
            // Source: Room.java:113-115
            this.resize(Random.NormalIntRange(minW, maxW) - 1,
                       Random.NormalIntRange(minH, maxH) - 1);
            return true;
        }
    }

    /**
     * Get a point n steps inside from a border point
     * Source: Room.java:120-132
     */
    pointInside(from, n) {
        const step = new Point(from);
        if (from.x === this.left) {
            step.offset(+n, 0);
        } else if (from.x === this.right) {
            step.offset(-n, 0);
        } else if (from.y === this.top) {
            step.offset(0, +n);
        } else if (from.y === this.bottom) {
            step.offset(0, -n);
        }
        return step;
    }

    /**
     * Width and height are increased by 1 because rooms are inclusive
     * Source: Room.java:134-143
     */
    width() {
        return super.width() + 1;
    }

    height() {
        return super.height() + 1;
    }

    /**
     * Random point inside room
     * Source: Room.java:145-152
     */
    random(m = 1) {
        return new Point(
            Random.IntRange(this.left + m, this.right - m + 1),
            Random.IntRange(this.top + m, this.bottom - m + 1)
        );
    }

    /**
     * Point is inside if within 1-tile perimeter
     * Source: Room.java:154-157
     */
    inside(p) {
        return p.x > this.left && p.y > this.top && p.x < this.right && p.y < this.bottom;
    }

    /**
     * Center with randomization for odd dimensions
     * Source: Room.java:159-163
     *
     * 🚨 CRITICAL: Uses Random.Int for odd-dimension randomization
     */
    center() {
        return new Point(
            Math.floor((this.left + this.right) / 2) + (((this.right - this.left) % 2) === 1 ? Random.Int_max(2) : 0),
            Math.floor((this.top + this.bottom) / 2) + (((this.bottom - this.top) % 2) === 1 ? Random.Int_max(2) : 0)
        );
    }

    // ========== Connection Logic ==========
    // Source: Room.java:166-290

    /**
     * Minimum connections for a direction
     * Source: Room.java:174-177
     */
    minConnections(direction) {
        if (direction === RoomDirection.ALL) return 1;
        else return 0;
    }

    /**
     * Current connections for a direction
     * Source: Room.java:179-194
     */
    curConnections(direction) {
        if (direction === RoomDirection.ALL) {
            return this.connected.size;
        } else {
            let total = 0;
            for (const r of this.connected.keys()) {
                const i = this.intersect(r);
                if (direction === RoomDirection.LEFT && i.width() === 0 && i.left === this.left) total++;
                else if (direction === RoomDirection.TOP && i.height() === 0 && i.top === this.top) total++;
                else if (direction === RoomDirection.RIGHT && i.width() === 0 && i.right === this.right) total++;
                else if (direction === RoomDirection.BOTTOM && i.height() === 0 && i.bottom === this.bottom) total++;
            }
            return total;
        }
    }

    /**
     * Remaining connections available
     * Source: Room.java:196-199
     */
    remConnections(direction) {
        if (this.curConnections(RoomDirection.ALL) >= this.maxConnections(RoomDirection.ALL)) return 0;
        else return this.maxConnections(direction) - this.curConnections(direction);
    }

    /**
     * Maximum connections for a direction
     * Source: Room.java:201-204
     */
    maxConnections(direction) {
        if (direction === RoomDirection.ALL) return 16;
        else return 4;
    }

    /**
     * Can connect at specific point?
     * Point must be along exactly one edge, no corners
     * Source: Room.java:207-210
     */
    canConnect_point(p) {
        return (p.x === this.left || p.x === this.right) !== (p.y === this.top || p.y === this.bottom);
    }

    /**
     * Can connect in direction?
     * Source: Room.java:213-215
     */
    canConnect_direction(direction) {
        return this.remConnections(direction) > 0;
    }

    /**
     * Can connect to another room?
     * Source: Room.java:218-245
     */
    canConnect_room(r) {
        // Entrance and exit rooms cannot directly connect
        // Source: Room.java:219-222
        if (this.isExit() && r.isEntrance() || this.isEntrance() && r.isExit()) {
            return false;
        }

        const i = this.intersect(r);

        // Must have at least one valid connection point
        let foundPoint = false;
        for (const p of i.getPoints()) {
            if (this.canConnect_point(p) && r.canConnect_point(p)) {
                foundPoint = true;
                break;
            }
        }
        if (!foundPoint) return false;

        // Check directional connection limits
        if (i.width() === 0 && i.left === this.left)
            return this.canConnect_direction(RoomDirection.LEFT) && r.canConnect_direction(RoomDirection.RIGHT);
        else if (i.height() === 0 && i.top === this.top)
            return this.canConnect_direction(RoomDirection.TOP) && r.canConnect_direction(RoomDirection.BOTTOM);
        else if (i.width() === 0 && i.right === this.right)
            return this.canConnect_direction(RoomDirection.RIGHT) && r.canConnect_direction(RoomDirection.LEFT);
        else if (i.height() === 0 && i.bottom === this.bottom)
            return this.canConnect_direction(RoomDirection.BOTTOM) && r.canConnect_direction(RoomDirection.TOP);
        else
            return false;
    }

    /**
     * Can merge with another room?
     * Source: Room.java:247-249
     */
    canMerge(level, other, p, mergeTerrain) {
        return false;
    }

    /**
     * Merge with another room
     * Source: Room.java:252-254
     */
    merge(level, other, merge, mergeTerrain) {
        // Will be implemented when Painter is ported
        // Painter.fill(level, merge, mergeTerrain);
    }

    /**
     * Add as neighbor
     * Rooms must share an edge of length >= 2
     * Source: Room.java:256-268
     */
    addNeigbour(other) {
        if (this.neigbours.includes(other)) {
            return true;
        }

        const i = this.intersect(other);
        if ((i.width() === 0 && i.height() >= 2) ||
            (i.height() === 0 && i.width() >= 2)) {
            this.neigbours.push(other);
            other.neigbours.push(this);
            return true;
        }
        return false;
    }

    /**
     * Connect to another room
     * Source: Room.java:270-279
     */
    connect(room) {
        if ((this.neigbours.includes(room) || this.addNeigbour(room))
                && !this.connected.has(room) && this.canConnect_room(room)) {
            this.connected.set(room, null);
            room.connected.set(this, null);
            return true;
        }
        return false;
    }

    /**
     * Clear all connections
     * Source: Room.java:281-290
     */
    clearConnections() {
        for (const r of this.neigbours) {
            r.neigbours = r.neigbours.filter(n => n !== this);
        }
        this.neigbours = [];

        for (const r of this.connected.keys()) {
            r.connected.delete(this);
        }
        this.connected.clear();
    }

    /**
     * Is this room an entrance?
     * Source: Room.java:292-294
     */
    isEntrance() {
        return false;
    }

    /**
     * Is this room an exit?
     * Source: Room.java:296-298
     */
    isExit() {
        return false;
    }

    // ========== Painter Logic ==========
    // Source: Room.java:300-382

    /**
     * Abstract paint method - must be overridden
     * Source: Room.java:302
     */
    paint(level) {
        throw new Error('Room.paint() must be overridden');
    }

    /**
     * Can place water at point?
     * Source: Room.java:305-307
     */
    canPlaceWater(p) {
        return true;
    }

    /**
     * Get all water-placeable points
     * Source: Room.java:309-318
     */
    waterPlaceablePoints() {
        const points = [];
        for (let i = this.left; i <= this.right; i++) {
            for (let j = this.top; j <= this.bottom; j++) {
                const p = new Point(i, j);
                if (this.canPlaceWater(p)) points.push(p);
            }
        }
        return points;
    }

    /**
     * Can place grass at point?
     * Source: Room.java:321-323
     */
    canPlaceGrass(p) {
        return true;
    }

    /**
     * Get all grass-placeable points
     * Source: Room.java:325-334
     */
    grassPlaceablePoints() {
        const points = [];
        for (let i = this.left; i <= this.right; i++) {
            for (let j = this.top; j <= this.bottom; j++) {
                const p = new Point(i, j);
                if (this.canPlaceGrass(p)) points.push(p);
            }
        }
        return points;
    }

    /**
     * Can place trap at point?
     * Source: Room.java:337-339
     */
    canPlaceTrap(p) {
        return true;
    }

    /**
     * Get all trap-placeable points
     * Source: Room.java:341-350
     */
    trapPlaceablePoints() {
        const points = [];
        for (let i = this.left; i <= this.right; i++) {
            for (let j = this.top; j <= this.bottom; j++) {
                const p = new Point(i, j);
                if (this.canPlaceTrap(p)) points.push(p);
            }
        }
        return points;
    }

    /**
     * Can place item at point?
     * Source: Room.java:353-355
     */
    canPlaceItem(p, level) {
        return this.inside(p);
    }

    /**
     * Get all item-placeable points
     * Source: Room.java:357-366
     */
    itemPlaceablePoints(level) {
        const points = [];
        for (let i = this.left; i <= this.right; i++) {
            for (let j = this.top; j <= this.bottom; j++) {
                const p = new Point(i, j);
                if (this.canPlaceItem(p, level)) points.push(p);
            }
        }
        return points;
    }

    /**
     * Can place character at point?
     * Source: Room.java:369-371
     */
    canPlaceCharacter(p, level) {
        return this.inside(p);
    }

    /**
     * Get all character-placeable points
     * Source: Room.java:373-382
     */
    charPlaceablePoints(level) {
        const points = [];
        for (let i = this.left; i <= this.right; i++) {
            for (let j = this.top; j <= this.bottom; j++) {
                const p = new Point(i, j);
                if (this.canPlaceCharacter(p, level)) points.push(p);
            }
        }
        return points;
    }

    // ========== Graph.Node Interface ==========
    // Source: Room.java:385-419

    /**
     * Get connected rooms (edges in graph)
     * Only includes unlocked/open doors
     * Source: Room.java:408-419
     */
    edges() {
        const edges = [];
        for (const [r, d] of this.connected.entries()) {
            // For path building, ignore locked, blocked, or hidden doors
            if (!d || d.type === DoorType.EMPTY || d.type === DoorType.TUNNEL
                    || d.type === DoorType.UNLOCKED || d.type === DoorType.REGULAR) {
                edges.push(r);
            }
        }
        return edges;
    }
}
